apps/examples/camera/camera_main.c

```diff
@@ -163,6 +193,149 @@ struct camera_ctrls ctrls[] = {
+	}
+	sem_init(&c->sem, 0, 1);
+}
+
+static int circ_bbuf_push(struct circ_buf *c, struct buf_object buf)
+{
+	sem_wait(&c->sem);
+	if ((c->head == 0 && c->tail == c->maxlen - 1) || (c->tail == (c->head - 1) % (c->maxlen - 1))) {
+		sem_post(&c->sem);
+		return -1;
+	}
+	/* Insert First Element */
+	else if (c->head == -1) {
+		c->head = c->tail = 0;
+		c->bobj[c->tail] = buf;
+	} else if (c->tail == c->maxlen - 1 && c->head != 0) {
+		c->tail = 0;
+		c->bobj[c->tail] = buf;
+	} else {
+		c->tail++;
+		c->bobj[c->tail] = buf;
+	}
+	sem_post(&c->sem);
+	return 0;
+}
+
+static int circ_bbuf_pop(struct circ_buf *c, struct buf_object **buf)
+{
+	sem_wait(&c->sem);
+	if (c->head == -1) {
+		sem_post(&c->sem);
+		return -1;
+	}
+	*buf = &c->bobj[c->head];
+	if (c->head == c->tail) {
+		c->head = -1;
+		c->tail = -1;
+	} else if (c->head == c->maxlen - 1) {
+		c->head = 0;
+	} else {
+		c->head++;
+	}
+	sem_post(&c->sem);
+	return 0;
+}
+
+static pthread_addr_t write_thread(pthread_addr_t arg)
+{
+	int ret = OK;
+	struct v4l2_buffer buf = { 0, };
+	struct write_context *pContext = (struct write_context *)arg;
+
+	memset(&buf, 0, sizeof(v4l2_buffer_t));
+	buf.memory = V4L2_MEMORY_USERPTR;
+
+	while (sem_wait(&pContext->sem) == 0) {
+		struct buf_object *bobject;
+		if (circ_bbuf_pop(&pContext->cbuf, &bobject) == -1) {
+			if (pContext->loop > 0) {
+				continue;
+			} else {
+				goto error;
+			}
+		}
+		write_file((uint8_t *) bobject->userptr, (size_t) bobject->bytesused, bobject->format, bobject->frame);
+		buf.type = bobject->buf_type;
+		buf.m.userptr = (unsigned long)bobject->userptr;
+		buf.length = bobject->length;
+		buf.index = bobject->index;
+		/* Note: VIDIOC_QBUF reset released buffer pointer. */
+		if (pContext->loop > 0) {
+			ret = ioctl(pContext->v_fd, VIDIOC_QBUF, (unsigned long)&buf);
+			if (ret) {
+				fprintf(stderr, "Fail QBUF %d\n", errno);
+				goto error;
+			}
+		} else {
+			goto error;
+		}
+	}
+error:
+	pthread_exit(NULL);
+	return NULL;
+}
+
+static pthread_t create_write_thread(void *arg)
+{
+	pthread_t tid;
+	pthread_attr_t attr;
+	struct sched_param sparam;
+	int ret = OK;
+
+	ret = pthread_attr_init(&attr);
+	if (ret != 0) {
+		fprintf(stderr, "failed to set pthread init(%d)\n", ret);
+		return -ret;
+	}
+
+	ret = pthread_attr_setschedpolicy(&attr, SCHED_RR);
+	if (ret != 0) {
+		fprintf(stderr, "failed to set policy(%d)\n", ret);
+	}
+
+	ret = pthread_attr_setstacksize(&attr, WRITE_THEAD_STACK_SIZE);
+	if (ret != 0) {
+		fprintf(stderr, "failed to set stack size(%d)\n", ret);
+		return -ret;
+	}
+
+	sparam.sched_priority = WRITE_THEAD_PRIORITY;
+	ret = pthread_attr_setschedparam(&attr, &sparam);
+	if (ret != 0) {
+		fprintf(stderr, "failed to set sched param(%d)\n", ret);
+		return -ret;
+	}
+
+	ret = pthread_create(&tid, &attr, write_thread, arg);
+	if (ret != 0) {
+		fprintf(stderr, "failed to create pthread(%d)\n", ret);
+		return -ret;
+	}
+
+	pthread_setname_np(tid, "write sigwait thread");
+	pthread_detach(tid);
```

Comment:
please check return value for success or failure..

Reply:
done